### ISSUE TITLE
Memory leak when calling `getMetadata()`

### ISSUE THREAD

**Issue by mrigger** [xerial/sqlite-jdbc#423]
Hi,

I discovered a memory leak that when calling `Connection#getMetaData()`.

Consider this example:

```
Connection con = DriverManager.getConnection("jdbc:sqlite:test.db");
while (true) {
	DatabaseMetaData meta = con.getMetaData();
	try (ResultSet tables = meta.getTables(null, null, null, null);) {
		// tables.getStatement().close();
	}
}
```
When executing this code, the JVM quickly runs out of memory, even though the `ResultSet` is closed. Using VisualVM, I found that most alive objects are `char[]`s and look like this:
```
char[]#1 : SELECT
  NULL AS TABLE_CAT,
  NULL AS TABLE_SCHEM,
  NAME AS TABLE_NAME,
  TYPE AS TABLE_TYPE,
  NULL AS REMARKS,
  NULL AS TYPE_CAT,
  NULL AS TYPE_SCHEM,
  NU...		1,512 B (0%)	n/a
```

Only when explicitly calling `tables.getStatement().close()` is the memory leak avoided.

**Comment by gotson** [comment]
Is this still happening on the latest version?

**Comment by mrigger** [comment]
I am unsure. I'm now using the workaround mentioned above. The test case above can be used to check whether the issue can still be reproduced.

**Comment by shabino** [comment]
I experienced this on the latest version. Unless the statement is manually closed, the connection retains a reference to the statement.

**Comment by gotson** [comment]
Just had a quick look, though i'm no expert in JDBC. I feel like the issue comes from [here](https://github.com/xerial/sqlite-jdbc/blob/e1d282cb2887ef427c7ad93d4fe7dd05a6c11473/src/main/java/org/sqlite/core/CoreResultSet.java#L124-L126).

When closing the ResultSet, the above will return, skipping the block where the statement is closed if it was marked to be closed.

Commenting out the block make my test pass, however it seems there's some history around this change: #358 and #215

@pyckle if you want to have a go at this one that would really help, seems like you have a good knowledge on concurrent access from the few PRs i've seen.

**Comment by pyckle** [comment]
From a brief look, I think you are right @gotson that it is from the check to see if the connection is open.

This check was added to avoid double closing result sets that resulted in correctness/performance issues for someone in #349.

The problem is that there are scenarios where open=false, and the statement has not yet been closed. I suspect the correct code is:

```
        if (stmt.pointer.isClosed() || (!open && !closeStmt)) {
            return;
        }
```

I don't have the time to test this code this week - work is busy. If I have free time sometime soon, I'll try and look more in depth.

**Comment by gotson** [comment]
@pyckle you rock, thanks 🚀

**Comment by pyckle** [comment]
I made a PR. let me know what you think.

The case of the query without anything returned seems to be handled properly already https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/DB.java#L975

But I added a unit test for this scenario nonetheless.

There as also a minor bugfix included for isAfterLast() - the jdbc spec says this should return false for a query without any rows, but it returned true.
